client/state/data-getters/index.js

```diff
@@ -209,3 +209,21 @@ export const requestSiteAlerts = siteId => {
 		}
 	);
 };
+
+export const requestAtomicSFTPDetails = siteId =>
+	requestHttpData(
+		`atomic-hosting-data-${ siteId }`,
+		http(
+			{
+				method: 'GET',
+				path: `/sites/${ siteId }/hosting/sftp`,
+				apiNamespace: 'wpcom/v2',
+			},
+			{}
+		),
+		{
+			fromApi: () => data => {
+				return [ 'sftpData', data ];
```

Comment:
this is great! Do we have a post where the expected format of this data is listed so that we can discuss it without diving into code?

once we have our expectations, I generally like to encourage people to make these `fromApi` functions even more explicit than we feel like they ought to. it serves as a way to get explicit data models in Calypso and to prevent "passing through" the opaque data structure. in other words, how do we know what properties of the API response need to be there? which are used? do we need any conversion on any of them (such as with dates)?

```js
fromApi: () => ( { username, password } ) => [
  [ `atomic-hosting-data-${ siteId }`, { username, password } ]
]
```

just a wee note there too - if we return `sftpData` as the key then it will conflate the SFTP details for _all sites_ with this value. typically we want to reuse the request ID and make sure that they are all globally-unique identifiers

---

sorry for the confusing bits of this `requestHttpData()` interface. you seem to have done a good job figuring it out but there are a few key oddities in there that do have specific reasons for being the way they are.

fun fact! `fromApi` takes a function that returns a function that parses the API response because many of our response parsers use `makeJSONSchemaParser()` and that's a noticeably-slow function to call. if we just passed it  directly without the outer thunk wrapper then we'd be calling that expensive function on every call to `render()`

fun fact! the return as an array of list pairs is there because in some API calls we get back information for more than one thing. it needs to be `[ [ id1, value1 ], [ id2, value2, … ]` and so technically this should fail the way it's written (it doesn't for another hidden gem in the system - the full unparsed API response is recorded in the request's id even though in most cases it should be replaced here.

imagine getting a list of posts - we might want to split the response into individual post item updates
```js
requestHttpData(
  `posts-for-site-${ siteId }`,
  http( … ),
  {
    fromApi: () => ( { posts } ) => posts.map( post => [
      post.id,
      post.post_content
    ] )
  }
);
```

Reply:
Thanks for this extra info, that explains a few things. Sorry about the delay in responding, I got tied up doing the API side of things and Olivia has taken over looking at the Redux end. 

The fromAPI return was definitely guess work on my part to get it working - so the background to this is really helpful!